ERCS/erc-7831.md

```diff
@@ -0,0 +1,367 @@
+
+<details>
+<summary>Python implementation of the modified ERC-55 algorithm</summary>
+
+```python
+from Crypto.Hash import keccak  # from pycryptodome
+
+
+def checksum_encode(addr):
+    hex_addr = addr.hex()
+    checksummed_buffer = ""
+
+    # Treat the hex address as ascii/utf-8 for keccak256 hashing
+    k = keccak.new(digest_bits=256)
+    k.update(hex_addr.encode("utf-8"))
+    hashed_address = k.hexdigest()
+
+    # Iterate over each character in the hex address
+    for nibble_index, character in enumerate(hex_addr):
+
+        if character in "0123456789":
+            # We can't upper-case the decimal digits
+            checksummed_buffer += character
+        elif character in "abcdef":
+            # Check if the corresponding hex digit (nibble) in the hash is 8 or higher
+            nibble_index_wrapped = nibble_index % len(hashed_address)
+            hashed_address_nibble = int(hashed_address[nibble_index_wrapped], 16)
+            if hashed_address_nibble > 7:
+                checksummed_buffer += character.upper()
+            else:
+                checksummed_buffer += character
+        else:
+            raise Exception(
+                f"Unrecognized hex character {character!r} at position {nibble_index}"
+            )
+
+    return "0x" + checksummed_buffer
+
+
+def test(addr_str: str):
+    padded_addr_str = addr_str.removeprefix("0x")
+    if len(padded_addr_str) % 2 == 1:
+        # Pad to an even number of nibbles.
+        padded_addr_str = "0" + padded_addr_str
+
+    addr_bytes = bytes.fromhex(padded_addr_str)
+    checksum_encoded = checksum_encode(addr_bytes)
+    if checksum_encoded != addr_str:
+        print(f"{checksum_encoded} != expected {addr_str}")
+
+
+test("0x5aAeb6053F3E94C9b9A09f33669435E7Ef1BeAed")
+test("0xfB6916095ca1df60bB79Ce92cE3Ea74c37c5d359")
+test("0xdbF03B407c01E7cD3CBea99509d93f8DDDC8C6FB")
+test("0xD1220A0cf47c7B9Be7A2E6BA89F429762e7b9aDb")
+test("0x004f67dAbb603AAA58eD52641CCafF09C559704A")
+test("0x4F67dABB603aAa58Ed52641cCAff09C559704A")
+test(
+    "0x"
+    "5aaeB6053f3e94c9B9a09f33669435E7"
+    "ef1BEaED5AaEB6053f3e94C9B9a09F33"
+    "669435e7ef1bEAed5aaEb6053f3e94C9"
+    "b9a09f33669435E7ef1beAED5aaEB605"
+    "3f3e94c9b9a09F33669435e7ef1beAED"
+)
+```
+
+</details>
+
+For example, these strings are correctly cased:
+
+* `004f67dAbb603AAA58eD52641CCafF09C559704A`
+*  `04f67dAbb603AAA58eD52641CCafF09C559704A`
+*   `4F67dABB603aAa58Ed52641cCAff09C559704A`
+
+[RFC 2119]: https://www.rfc-editor.org/rfc/rfc2119
+[RFC 8174]: https://www.rfc-editor.org/rfc/rfc8174
+[RFC 5234]: https://www.rfc-editor.org/rfc/rfc5234
+[RFC 7405]: https://www.rfc-editor.org/rfc/rfc7405
+[ERC-55]: ./eip-55.md
+[ERC-137]: ./eip-137.md
+[EIP-155]: ./eip-155.md
+[ERC-165]: ./eip-165.md
+[ERC-2304]: ./eip-2304.md
+
+## Rationale
+
+### Component Order
+
+The components are ordered from most specific to most general because... <!-- TODO -->
```

Comment:
so the question being answered here is: why address first and then chain specification next instead of other way around where we have been writing chain:address format